File: python/ch10/pipeline.py
# python/ch10/pipeline.py
"""貪欲 → 山登り → 焼きなましの多段パイプライン（第10章）."""
from __future__ import annotations

import math
import random
import sys
import time
from dataclasses import dataclass
from typing import List, TextIO

TOTAL_SEC = 2.0
CONSTRUCT_RATIO = 0.6

T0 = 5000.0
ALPHA = 0.995


@dataclass
class ProblemInput:
    days: int
    decay: List[int]
    gain: List[List[int]]


@dataclass
class ScheduleState:
    types: List[int]

    def copy(self) -> ScheduleState:
        return ScheduleState(list(self.types))


def read_problem(stream: TextIO) -> ProblemInput:
    lines = stream.read().splitlines()
    idx = 0
    days = int(lines[idx])
    idx += 1
    decay = list(map(int, lines[idx].split()))
    idx += 1
    gain = []
    for _ in range(days):
        gain.append(list(map(int, lines[idx].split())))
        idx += 1
    return ProblemInput(days=days, decay=decay, gain=gain)


def score_schedule(problem: ProblemInput, state: ScheduleState) -> int:
    last = [0] * 26
    satisfaction = 0
    for day in range(1, problem.days + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
    return max(10**6 + satisfaction, 0)


def greedy_schedule(problem: ProblemInput) -> ScheduleState:
    schedule: List[int] = []
    for day in range(1, problem.days + 1):
        best_t, best_sat = 1, None
        for t in range(1, 27):
            trial = ScheduleState(schedule + [t])
            sat = _final_satisfaction(problem, trial)
            if best_sat is None or sat > best_sat:
                best_sat = sat
                best_t = t
        schedule.append(best_t)
    return ScheduleState(schedule)


def _final_satisfaction(problem: ProblemInput, state: ScheduleState) -> int:
    last = [0] * 26
    satisfaction = 0
    for day in range(1, len(state.types) + 1):
        t = state.types[day - 1] - 1
        satisfaction += problem.gain[day - 1][t]
        last[t] = day
        decay = sum(problem.decay[i] * (day - last[i]) for i in range(26))
        satisfaction -= decay
    return satisfaction


def hill_climb(
    problem: ProblemInput,
    state: ScheduleState,
    deadline: float,
    rng: random.Random,
) -> ScheduleState:
    current = state.copy()
    score = score_schedule(problem, current)
    steps = 0
    while time.perf_counter() < deadline:
        day = rng.randrange(problem.days)
        old_type = current.types[day]
        improved = False
        for t in range(1, 27):
            if t == old_type:
                continue
            trial = current.copy()
            trial.types[day] = t
            trial_score = score_schedule(problem, trial)
            if trial_score > score:
                current = trial
                score = trial_score
                improved = True
                break
        steps += 1
        if steps % 200 == 0:
            print(f"hc step={steps} score={score}", file=sys.stderr)
        if not improved and steps % 50 == 0:
            continue
    return current


def sa(
    problem: ProblemInput,
    state: ScheduleState,
    deadline: float,
    rng: random.Random,
) -> ScheduleState:
    current = state.copy()
    score = score_schedule(problem, current)
    best = current.copy()
    best_score = score
    temperature = T0
    step = 0

    while time.perf_counter() < deadline:
        day = rng.randrange(problem.days)
        trial = current.copy()
        trial.types[day] = rng.randint(1, 26)
        trial_score = score_schedule(problem, trial)
        delta = trial_score - score
        if delta >= 0 or rng.random() < math.exp(delta / temperature):
            current = trial
            score = trial_score
            if score > best_score:
                best = current.copy()
                best_score = score
        temperature *= ALPHA
        step += 1
        if step % 500 == 0:
            print(f"sa step={step} score={score} T={temperature:.2f}", file=sys.stderr)

    return best


def main() -> None:
    rng = random.Random(42)
    problem = read_problem(sys.stdin)
    t0 = time.perf_counter()
    end = t0 + TOTAL_SEC
    construct_end = t0 + TOTAL_SEC * CONSTRUCT_RATIO

    print("phase=greedy start", file=sys.stderr)
    state = greedy_schedule(problem)
    print(f"phase=greedy end score={score_schedule(problem, state)}", file=sys.stderr)

    print("phase=hill_climb start", file=sys.stderr)
    state = hill_climb(problem, state, construct_end, rng)
    print(f"phase=hill_climb end score={score_schedule(problem, state)}", file=sys.stderr)

    print("phase=sa start", file=sys.stderr)
    state = sa(problem, state, end, rng)
    final = score_schedule(problem, state)
    print(f"phase=sa end score={final}", file=sys.stderr)
    print(f"elapsed={time.perf_counter()-t0:.3f}s", file=sys.stderr)

    for t in state.types:
        print(t)


if __name__ == "__main__":
    main()

File: go/ch10/pipeline.go
// go/ch10/pipeline.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	totalSec        = 2.0
	constructRatio  = 0.6
	t0              = 5000.0
	alpha           = 0.995
)

type ProblemInput struct {
	Days  int
	Decay []int
	Gain  [][]int
}

type ScheduleState struct {
	Types []int
}

func readProblem(r *bufio.Reader) (ProblemInput, error) {
	line, err := r.ReadString('\n')
	if err != nil {
		return ProblemInput{}, err
	}
	days, _ := strconv.Atoi(strings.TrimSpace(line))
	line, err = r.ReadString('\n')
	if err != nil {
		return ProblemInput{}, err
	}
	decay := parseInts(strings.TrimSpace(line))
	gain := make([][]int, days)
	for i := 0; i < days; i++ {
		line, err = r.ReadString('\n')
		if err != nil {
			return ProblemInput{}, err
		}
		gain[i] = parseInts(strings.TrimSpace(line))
	}
	return ProblemInput{days, decay, gain}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func copyState(s ScheduleState) ScheduleState {
	return ScheduleState{append([]int{}, s.Types...)}
}

func scoreSchedule(problem ProblemInput, state ScheduleState) int {
	last := make([]int, 26)
	satisfaction := 0
	for day := 1; day <= problem.Days; day++ {
		t := state.Types[day-1] - 1
		satisfaction += problem.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += problem.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
	}
	score := 1000000 + satisfaction
	if score < 0 {
		return 0
	}
	return score
}

func finalSatisfaction(problem ProblemInput, state ScheduleState) int {
	last := make([]int, 26)
	satisfaction := 0
	for day := 1; day <= len(state.Types); day++ {
		t := state.Types[day-1] - 1
		satisfaction += problem.Gain[day-1][t]
		last[t] = day
		decay := 0
		for i := 0; i < 26; i++ {
			decay += problem.Decay[i] * (day - last[i])
		}
		satisfaction -= decay
	}
	return satisfaction
}

func greedySchedule(problem ProblemInput) ScheduleState {
	schedule := make([]int, 0, problem.Days)
	for day := 1; day <= problem.Days; day++ {
		bestT, bestSat := 1, int(-1e18)
		for t := 1; t <= 26; t++ {
			trial := append(append([]int{}, schedule...), t)
			sat := finalSatisfaction(problem, ScheduleState{trial})
			if sat > bestSat {
				bestSat = sat
				bestT = t
			}
		}
		schedule = append(schedule, bestT)
	}
	return ScheduleState{schedule}
}

func hillClimb(problem ProblemInput, state ScheduleState, deadline time.Time, rng *rand.Rand) ScheduleState {
	current := copyState(state)
	score := scoreSchedule(problem, current)
	steps := 0
	for time.Now().Before(deadline) {
		day := rng.Intn(problem.Days)
		oldType := current.Types[day]
		improved := false
		for t := 1; t <= 26; t++ {
			if t == oldType {
				continue
			}
			trial := copyState(current)
			trial.Types[day] = t
			trialScore := scoreSchedule(problem, trial)
			if trialScore > score {
				current = trial
				score = trialScore
				improved = true
				break
			}
		}
		steps++
		if steps%5000 == 0 {
			fmt.Fprintf(os.Stderr, "hc step=%d score=%d\n", steps, score)
		}
		_ = improved
	}
	return current
}

func sa(problem ProblemInput, state ScheduleState, deadline time.Time, rng *rand.Rand) ScheduleState {
	current := copyState(state)
	score := scoreSchedule(problem, current)
	best := copyState(current)
	bestScore := score
	temperature := t0
	step := 0

	for time.Now().Before(deadline) {
		day := rng.Intn(problem.Days)
		trial := copyState(current)
		trial.Types[day] = rng.Intn(26) + 1
		trialScore := scoreSchedule(problem, trial)
		delta := float64(trialScore - score)
		if delta >= 0 || rng.Float64() < math.Exp(delta/temperature) {
			current = trial
			score = trialScore
			if score > bestScore {
				best = copyState(current)
				bestScore = score
			}
		}
		temperature *= alpha
		step++
		if step%5000 == 0 {
			fmt.Fprintf(os.Stderr, "sa step=%d score=%d T=%.2f\n", step, score, temperature)
		}
	}
	return best
}

func main() {
	rng := rand.New(rand.NewSource(42))
	problem, err := readProblem(bufio.NewReader(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	start := time.Now()
	end := start.Add(time.Duration(totalSec * float64(time.Second)))
	constructEnd := start.Add(time.Duration(totalSec * constructRatio * float64(time.Second)))

	fmt.Fprintln(os.Stderr, "phase=greedy start")
	state := greedySchedule(problem)
	fmt.Fprintf(os.Stderr, "phase=greedy end score=%d\n", scoreSchedule(problem, state))

	fmt.Fprintln(os.Stderr, "phase=hill_climb start")
	state = hillClimb(problem, state, constructEnd, rng)
	fmt.Fprintf(os.Stderr, "phase=hill_climb end score=%d\n", scoreSchedule(problem, state))

	fmt.Fprintln(os.Stderr, "phase=sa start")
	state = sa(problem, state, end, rng)
	final := scoreSchedule(problem, state)
	fmt.Fprintf(os.Stderr, "phase=sa end score=%d\n", final)
	fmt.Fprintf(os.Stderr, "elapsed=%.3fs\n", time.Since(start).Seconds())

	for _, t := range state.Types {
		fmt.Println(t)
	}
}
